modules/imgcodecs/src/grfmt_gdal.cpp

```diff
@@ -388,38 +388,66 @@ bool GdalDecoder::readData( Mat& img ){
         // get the GDAL Band
         GDALRasterBand* band = m_dataset->GetRasterBand(c+1);
 
-        // make sure the image band has the same dimensions as the image
-        if( band->GetXSize() != m_width || band->GetYSize() != m_height ){ return false; }
+        /* Map palette band and gray band to color index 0 and red, green,
+           blue, alpha bands to BGRA indexes. Note: ignoring HSL, CMY,
+           CMYK, and YCbCr color spaces, rather than converting them
+           to BGR. */
+        int color = -1;
+        switch (band->GetColorInterpretation()) {
+        case GCI_PaletteIndex:
+        case GCI_GrayIndex:
+        case GCI_BlueBand:
+            color = 0;
+            break;
+        case GCI_GreenBand:
+            color = 1;
+            break;
+        case GCI_RedBand:
+            color = 2;
+            break;
+        case GCI_AlphaBand:
+            color = 3;
+            break;
+        default:
+            color = -1;
+            break;
```

Comment:
Probably we should replace code to this in default case:
```
CV_Error(Error::StsNotImplemented, "unsupported/invalid image format");
```

Also we could remove `if (color >= 0)` below.

Reply:
That sounds like a good idea to me.

Reply:
I've pushed this update. Let me know if it is what you had in mind.